bgpd/bgpd.c

```diff
@@ -3809,8 +3813,20 @@ peer_flag_modify_action (struct peer *peer, u_int32_t flag)
             peer_nsf_stop (peer);
 
 	  if (BGP_IS_VALID_STATE_FOR_NOTIF(peer->status))
-	    bgp_notify_send (peer, BGP_NOTIFY_CEASE,
-			     BGP_NOTIFY_CEASE_ADMIN_SHUTDOWN);
+            {
+              char *msg = peer->tx_shutdown_message;
+              size_t msglen;
+
+              if (!msg && peer_group_active (peer))
+                 msg = peer->group->conf->tx_shutdown_message;
+              msglen = msg ? strlen(msg) : 0;
+              if (msglen > 128)
+                 msglen = 128;
+
+              bgp_notify_send_with_data (peer, BGP_NOTIFY_CEASE,
+                                         BGP_NOTIFY_CEASE_ADMIN_SHUTDOWN,
+                                         (uint8_t *)msg, msglen);
```

Comment:
Yea, I think the code doesn't implement the specified protocol correctly, as it is. Working on it. :)